### PR TITLE
VAULT-19255 - Add event based static secret cache updater to Vault Proxy

### PR DESCRIPTION
**Pull request by VioletHynes**
A follow-up to #23177 that adds a new subsystem to Vault Proxy that, when static secret caching is enabled, will update the cache upon received events. A new config validation has been added to ensure that an auto-auth token is in use alongside `cache_static_secrets`. I also fixed a bug where KV Puts were being cached -- whoops!

This pull request is quite large, but a significant amount of it is tests. I've added a _lot_ of tests at all levels for this, functionality, including some sanity checks for the event system that weren't tested elsewhere to my knowledge.

What's still missing:
The updating of a token's permissions via /sys/capabilities. This will come in a later PR and is part of a later ticket.

I've set no changelog for this, as I'll do a large one as a feature changelog near the end.

### CODE REVIEW

File: vault/testing.go
Comment (VioletHynes) [inline vault/testing.go:2149]: I did this opportunistic refactor as part of investigating why the event system doesn't seem to send events for the default KVv1 mount, as this code was copy/pasted.

File: command/agentproxyshared/cache/lease_cache.go
Comment (VioletHynes) [inline command/agentproxyshared/cache/lease_cache.go:212]: If we invalidate the cache on a PUT, there's no cached entry to update once the event comes through with details about the PUT.

**Review by kubawi — APPROVED** [review]
LGTM! 
Thanks for the detailed PR description and all the code comments, they made the review much smoother! 

I've made sure not to rush this review, but given the scale of the changes and that it's a subsystem I don't have a lot of experience with I would recommend waiting for another review before merging 😄 

PS I've used [Josh's PR Review comment key](https://hashicorp.atlassian.net/wiki/spaces/VAULT/pages/2545254986/PR+Review+comment+key) in the review.

File: command/agentproxyshared/cache/lease_cache.go
Comment (VioletHynes) [reply command/agentproxyshared/cache/lease_cache.go:209]: Good shout! Let me update all of these :)

File: command/agentproxyshared/cache/lease_cache.go
Comment (VioletHynes) [reply command/agentproxyshared/cache/lease_cache.go:207]: We could, but it makes more sense to me for `checkCacheForRequest` to take a request, rather than the token of that request. I also think it's not unlikely we could want other aspects of the `req` (such as logging the path) in the future!

File: command/agentproxyshared/cache/lease_cache.go
Comment (VioletHynes) [reply command/agentproxyshared/cache/lease_cache.go:224]: Token could be empty here, for a few reasons:
- Even though we need auto-auth for static secret caching, there's no guarantee that `use_auto_auth_token` is set on the `api_proxy` setting
- The request could be to an unauthenticated endpoint

I'll leave a comment to make this clearer, though!

File: command/agentproxyshared/cache/static_secret_cache_updater.go
Comment (VioletHynes) [reply command/agentproxyshared/cache/static_secret_cache_updater.go:82]: Good point! This was a hold-over from a copy/paste, I think

File: command/agentproxyshared/cache/static_secret_cache_updater.go
Comment (VioletHynes) [reply command/agentproxyshared/cache/static_secret_cache_updater.go:135]: I think the existing message is clearer, as while this is about a specific part of the event, ultimately, something has gone very wrong if the event is in the wrong shape. In particular, we'd get this message if there was no `data` element in the JSON, so data format isn't wrong, it's not present :)

File: command/agentproxyshared/cache/static_secret_cache_updater.go
Comment (VioletHynes) [reply command/agentproxyshared/cache/static_secret_cache_updater.go:155]: Good catch! This was a copy/paste error I think

File: command/agentproxyshared/cache/static_secret_cache_updater.go
Comment (VioletHynes) [reply command/agentproxyshared/cache/static_secret_cache_updater.go:162]: We don't, though I thought it added to the code clarity. If it makes things more confusing in your opinion, I'm happy to remove it :)

File: command/agentproxyshared/cache/static_secret_cache_updater.go
Comment (VioletHynes) [reply command/agentproxyshared/cache/static_secret_cache_updater.go:120]: We do! `ctx.Done()` happens if the context is cancelled or timed out.

File: command/agentproxyshared/cache/static_secret_cache_updater.go
Comment (VioletHynes) [reply command/agentproxyshared/cache/static_secret_cache_updater.go:181]: Fixed! Thanks :)

File: command/agentproxyshared/cache/static_secret_cache_updater.go
Comment (VioletHynes) [reply command/agentproxyshared/cache/static_secret_cache_updater.go:204]: `RawRequestWithContext` is different, yeah. It's deprecated, but as far as I can tell it's not so much deprecated as it is trying to encourage you to use higher level functions, which are better 99.9% of the time, we're just in that 0.1%. It's still used in production code (like in our API proxy). This also matches the same way we send the requests in the API proxy.

It looks like `ReadRawWithContext` was added earlier this year for a similar use case, but it's still calling `RawRequestWithContext` under the hood. I don't really know that there's much of a difference other than the fact it creates the `Request` in the function. To put it another way, they're both using the same 'deprecated' path, but I think deprecated can be a bit misleading here.

File: command/agentproxyshared/cache/static_secret_cache_updater.go
Comment (VioletHynes) [reply command/agentproxyshared/cache/static_secret_cache_updater.go:294]: Good call!

File: command/agentproxyshared/cache/static_secret_cache_updater.go
Comment (VioletHynes) [reply command/agentproxyshared/cache/static_secret_cache_updater.go:332]: Good shout! Thanks :)

File: command/proxy/config/config.go
Comment (VioletHynes) [reply command/proxy/config/config.go:253]: We do! This is the new test in `config_test.go` :)

File: vault/testing.go
Comment (VioletHynes) [reply vault/testing.go]: I removed this from this PR, and moved it into https://github.com/hashicorp/vault/pull/23584 - so resolving this conversation

File: command/agentproxyshared/cache/static_secret_cache_updater_test.go
Comment (VioletHynes) [reply command/agentproxyshared/cache/static_secret_cache_updater_test.go:139]: Good shout! Updated :)

File: command/agentproxyshared/cache/static_secret_cache_updater_test.go
Comment (VioletHynes) [reply command/agentproxyshared/cache/static_secret_cache_updater_test.go:1]: I'm open to ideas, if you have any :)

File: command/agentproxyshared/cache/static_secret_cache_updater.go
Comment (VioletHynes) [reply command/agentproxyshared/cache/static_secret_cache_updater.go]: To put it another way, if we receive `{ "foo": "bar"}` as our event, the event format is wrong, as opposed to the format of the (missing) `data` object. There are no valid events without the `data` object, so something has gone quite wrong with the event if this ever happens.

I believe this applies to the other, similar comments you had too, but let me know if you think it doesn't!

Comment (kubawi) [reply command/agentproxyshared/cache/static_secret_cache_updater.go]: What worried me was that we have four identical error message strings in the code which might make pinpointing the exact code path when debugging more difficult, but I guess maybe in that case the value of `message` could point us in the right direction? 

In general, I like when error message strings are as unique as possible to allow searching the codebase during incidents, etc. simpler, but it might be an abundance of caution from my side 😄

File: command/agentproxyshared/cache/static_secret_cache_updater.go
Comment (kubawi) [reply command/agentproxyshared/cache/static_secret_cache_updater.go]: I guess that during the review I did a double take to make sure that there wasn't any missing code there, but on deeper thought that's not how we would read that code coming back to it. Either way, I think the comment makes the intention clear, so maybe I was just overzealous here, soz! 😄

File: command/agentproxyshared/cache/static_secret_cache_updater.go
Comment (VioletHynes) [reply command/agentproxyshared/cache/static_secret_cache_updater.go]: It should (hopefully) be immediately obvious from the error what's wrong, but you're right, knowing where the error came from could definitely help! Let me update to e.g. `unexpected event format when decoding 'data' element`, does that work? :)

Comment (kubawi) [reply command/agentproxyshared/cache/static_secret_cache_updater.go]: Yeah, that sounds spot on. Thanks, Violet!

Comment (VioletHynes) [reply command/agentproxyshared/cache/static_secret_cache_updater.go]: Updated! Thanks a bunch! :)

**Pull request merged**
